clusterversion: document process for introducing cluster versions

irfansharif opened the pull request:
And for deleting old ones. This stuff is endlessly confusing.

Release note: None

---

(I could be wrong about some of it, do scrutinize.)

irfansharif committed:
clusterversion: document process for introducing cluster versions

And for deleting old ones. This stuff is endlessly confusing.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57154)
<!-- Reviewable:end -->

irfansharif committed:
clusterversion: move version comments next to definition

Jump to symbol for a given version tag will now group together the
intent of the version itself (instead of expecting readers to know to
grep for the same version in the block below).

Release note: None

tbg reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-:-MN7vkqi93s7VD4rIp-B:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @irfansharif, and @lucy-zhang)

___
*[pkg/clusterversion/cockroach_versions.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-MN7tcpLEze5r8pfiX9g:-MN7tcpLEze5r8pfiX9h:b-8mummo) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8e7df198154e4cfb6eae96b838fe4c71bc506f/pkg/clusterversion/cockroach_versions.go#L35)):*
> ```Go
> //
> //     See clusterversion.Handle.IsActive and usage of existing versions below
> //     for additional commentary.
> ```

This could be a good place to highlight an "anti-pattern". Basically, a node should not refuse functionality it can handle on the grounds that the version is *not* active. The initiator of a request gets to decide what is supported. For example, this here is bad:

```
func somewhereInSQL() {
  if (version is active) { // <-- this is good
    req.NewFeatureFlag = true
  }
  send(req)
}
```

```
func handleReq(req Req) error {
  if req.NewFeatureFlag && !(version is active) { // <-- this is bad and also unnecessary
    return errors.New("version not active!")
  }
  // ...
}
```

However, we need to prevent end-users from enabling features before they're supported, so this is fine:

```
func handleSomeNewStatement() error {
  if !(version is active) {
    return errors.New("cannot use NewStatement until cluster is fully upgraded")
  }
}
```
___
*[pkg/clusterversion/cockroach_versions.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-MN7uds7EvnCBNggg1gx:-MN7uds7EvnCBNggg1gy:bctc7fl) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8e7df198154e4cfb6eae96b838fe4c71bc506f/pkg/clusterversion/cockroach_versions.go#L41)):*
> ```Go
> //
> //       (i)  Version20_2 (keyed to v20.2.0-0})
> //	     (ii) VersionStart21_1 (keyed to v20.2.0-1})
> ```

weird indent. Also, is the -0 and -1 going to rot with your next PR?
___
*[pkg/clusterversion/cockroach_versions.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-MN7v3gf2Ct0PooIB3Ib:-MN7uwikE43TRSMr3ICA:b9hfb4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8e7df198154e4cfb6eae96b838fe4c71bc506f/pkg/clusterversion/cockroach_versions.go#L47)):*
> ```Go
> //    example above, you'll set it to Version20_2. This indicates that the
> //    minimum binary version required in a cluster with with nodes running
> //    v21.1 binaries (including pre-release alphas) is v20.2
> ```

, i.e. that an upgrade into such a binary must start out from at least v20.2 nodes.
At the time of writing, the binary min supported version is the last major release, though we may consider relaxing this in the future (i.e. for example could skip up to one major release) as we move to a more frequent release schedule.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-:-MNAgUg_5IjAq0UdEQUx:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @tbg)

___
*[pkg/clusterversion/cockroach_versions.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-MN7tcpLEze5r8pfiX9g:-MNAgWrt2xl3PMac0JO5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8e7df198154e4cfb6eae96b838fe4c71bc506f/pkg/clusterversion/cockroach_versions.go#L35)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This could be a good place to highlight an "anti-pattern". Basically, a node should not refuse functionality it can handle on the grounds that the version is *not* active. The initiator of a request gets to decide what is supported. For example, this here is bad:

```
func somewhereInSQL() {
  if (version is active) { // <-- this is good
    req.NewFeatureFlag = true
  }
  send(req)
}
```

```
func handleReq(req Req) error {
  if req.NewFeatureFlag && !(version is active) { // <-- this is bad and also unnecessary
    return errors.New("version not active!")
  }
  // ...
}
```

However, we need to prevent end-users from enabling features before they're supported, so this is fine:

```
func handleSomeNewStatement() error {
  if !(version is active) {
    return errors.New("cannot use NewStatement until cluster is fully upgraded")
  }
}
```

</blockquote></details>

Done.
___
*[pkg/clusterversion/cockroach_versions.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-MN7uds7EvnCBNggg1gx:-MNAgYN7CwcnLHWwcoaM:b-rxavk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8e7df198154e4cfb6eae96b838fe4c71bc506f/pkg/clusterversion/cockroach_versions.go#L41)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

weird indent. Also, is the -0 and -1 going to rot with your next PR?
</blockquote></details>

Done. And kind of, but I'll leave things as is. I'm referring to the internal versions directly here vs. the printed form, and these version keys will have to be defined per-release by hand.
___
*[pkg/clusterversion/cockroach_versions.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57154#-MN7v3gf2Ct0PooIB3Ib:-MNAgY8T7TG28bPlTtFI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8e7df198154e4cfb6eae96b838fe4c71bc506f/pkg/clusterversion/cockroach_versions.go#L47)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

, i.e. that an upgrade into such a binary must start out from at least v20.2 nodes.
At the time of writing, the binary min supported version is the last major release, though we may consider relaxing this in the future (i.e. for example could skip up to one major release) as we move to a more frequent release schedule.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

irfansharif commented:
bors r+
